Fix regexp error message in dictionaries

CyrilleB79 opened the pull request:
### Link to issue number:
Fixes #14556
### Summary of the issue:

Dictionaries do not allow to use backslash character normally in the replacement field when regexp is not selected as entry type. When such a character is used, a message box complaining of a regexp error appears.

### Description of user facing changes

* The user can use backslash in replacement field normally when entry type is not regexp
* Error messages are more precise and more specific when entering erroneous values in the fields for a regexp entry

### Description of development approach
* When the entry type is not regexp, escape the replacement string
* When entries are created:
  * filter exception on `re.error` instead of the too global `Exception` to catch only exceptions related to regexps
  * separate the try/except statement in two in order to indicate in which field the regexp error lies
  * In case of a regexp error, after validating "OK" in the message box, the field where the error lies is re-focused
  * Reports the regexp error only in case of entry type = regexp; in the other cases,re-raise the exception, in order not to catch unrelated exceptions, e.g. code bug.

### Testing strategy:

Manual tests:

* Add entry with backslash in replacement field and entry type = everywhere
* Add entry with error either in pattern or in replacement field with entry type = regexp.


### Known issues with pull request:
None
### Change log entries:
Bug fixes
`It is now possible to use the backslash character in the replacement field of a dictionaries entry, when the type is not set to regular expression. (#14556)`

### Code Review Checklist:

 [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

CyrilleB79 committed:
Fix regexp error message in dictionaries.

XLTechie commented:
For changelog:

> It is now again possible to use backslash character in the replacement field of dictionaries. (#14556)

Suggestion:

It is now possible to use the backslash character in the replacement field of dictionaries, when the pattern is not set to regular expression. (#14556)

XLTechie commented on source/gui/speechDict.py:
```diff
@@ -94,20 +96,37 @@ def onOk(self, evt):
-		except Exception as e:
-			log.debugWarning("Could not add dictionary entry due to (regex error) : %s" % e)
+		except RegexpError as e:
+			log.debugWarning("Could not add dictionary entry due to (regex error in the replacement field) : %s" % e)
```
Why keep the parentheses around the error itself? "(regex error in the replacement field)"

And perhaps I am missing something, but why not switch to F-strings for these messages?

CyrilleB79 committed:
Review actions: slight reformatting of messages.

CyrilleB79 commented:
@XLTechie thanks for the review. I have addressed the raised points in the code and in the initial description.

seanbudd commented on source/speechDictHandler/__init__.py:
```diff
@@ -1,8 +1,8 @@
-#Copyright (C) 2006-2017 NVDA Contributors <http://www.nvda-project.org/>
-#This file is covered by the GNU General Public License.
-#See the file COPYING for more details.
+# speechDictHandler.py
```
```suggestion
```

seanbudd commented on source/gui/speechDict.py:
```diff
@@ -94,20 +96,37 @@ def onOk(self, evt):
 			gui.messageBox(
 				# Translators: This is an error message to let the user know that the dictionary entry is not valid.
-				_("Regular Expression error: \"%s\".") % e,
+				_("Regular Expression error in the replacement field: \"{e}\"."),
```
Missing f in fstring

```suggestion
				_(f"Regular Expression error in the replacement field: \"{e}\"."),
```

seanbudd reviewed (commented):
Thanks @CyrilleB79

CyrilleB79 committed:
Update source/speechDictHandler/__init__.py

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

CyrilleB79 committed:
Do not use f-strings with gettext

CyrilleB79 committed:
Focus the field in which there is an error.

CyrilleB79 commented:
@seanbudd, I have addressed review comments. However, it is not possible to use f-strings with gettext so I have turned back to % interpolation as it was before.

I have also added two more lines to focus the field causing the regexp error as it was already done in case of empty pattern field error.

XLTechie commented:
> However, it is not possible to use f-strings with gettext so I have turned back to % interpolation as it was before.

I meant only to suggest that for the log messages, not the user messages. 
Apologies if I commented on the wrong line.

XLTechie commented:
> However, it is not possible to use f-strings with gettext so I have turned back to % interpolation as it was before.

It is possible, however, to use .format() with gettext, if you want to go 
slightly more modern than % replacements in the user messages. See 
globalCommands.py for examples, among others.

CyrilleB79 commented:
> > However, it is not possible to use f-strings with gettext so I have turned back to % interpolation as it was before.
> I meant only to suggest that for the log messages, not the user messages. Apologies if I commented on the wrong line.

No worry, you commented on the correct line. But I have then added another extra f-string in an inappropriate place, i.e. where gettext is used.

Regarding `%` versus `.format()` syntax, is there a recommendation or a preference to use the latter? (NVDA/NVAccess recommendation and/or Python/PEP recommendation)
 I'll let @seanbudd  decide if a change should be done.

josephsl commented:
Hi, regarding formatted string literals (F strings) and string formatting, the former is not used in translatable strings due to security (translated messages can introduce strings that could crash Python), and using str.format allows translators to see more readable (and understandable) source text for translations. Thanks.

seanbudd commented:
Agreed with @josephsl, the label in the format string e.g. `{settingsName}` is more explicit and helpful than `%s`

CyrilleB79 committed:
Reformat error message strings.

CyrilleB79 commented:
> Agreed with @josephsl, the label in the format string e.g. `{settingsName}` is more explicit and helpful than `%s`

OK @seanbudd, I have fixed it.

seanbudd commented on source/gui/speechDict.py:
```diff
@@ -94,25 +96,44 @@ def onOk(self, evt):
+		except RegexpError as e:
+			log.debugWarning(f"Could not add dictionary entry due to regex error in the pattern field : {e}")
+			if entryType != speechDictHandler.ENTRY_TYPE_REGEXP:
+				raise(e)
```
```suggestion
				raise e
```

seanbudd commented on source/gui/speechDict.py:
```diff
@@ -94,25 +96,44 @@ def onOk(self, evt):
+		except RegexpError as e:
+			log.debugWarning(f"Could not add dictionary entry due to regex error in the replacement field : {e}")
+			if entryType != speechDictHandler.ENTRY_TYPE_REGEXP:
+				raise(e)
```
```suggestion
				raise e
```

seanbudd reviewed (commented):
Looks good to me, other than a minor syntax issue

seanbudd committed:
Apply suggestions from code review

seanbudd commented on source/speechDictHandler/__init__.py:
```diff
@@ -42,7 +41,10 @@ def __init__(self, pattern, replacement,comment,caseSensitive=True,type=ENTRY_TY
 		self.type=type
 
 	def sub(self, text):
```
```suggestion
	def sub(self, text: str) -> str:
```

seanbudd committed:
add typing

seanbudd committed:
Merge remote-tracking branch 'origin/master' into regexpError

seanbudd committed:
update changes
